rpm/packagescanner.go

```diff
@@ -144,133 +96,113 @@ func (ps *Scanner) Scan(ctx context.Context, layer *claircore.Layer) ([]*clairco
 		}
 		done[db.Path] = struct{}{}
 
+		var nat nativeDB // see native_db.go:/nativeDB
 		switch db.Kind {
-		case kindBDB:
-			// Using --root and --dbpath, run rpm query on every suspected database.
-			// RPM interprets the absolute path passed to "dbpath" as underneath "root".
-			cmd := exec.CommandContext(ctx, "rpm",
-				`--root`, extractRoot, `--dbpath`, filepath.Join("/", db.Path),
-				`--query`, `--all`, `--queryformat`, queryFmt)
-			r, err := cmd.StdoutPipe()
+		case kindSQLite:
+			r, err := sys.Open(path.Join(db.Path, `rpmdb.sqlite`))
 			if err != nil {
-				return nil, err
+				return nil, fmt.Errorf("rpm: error reading sqlite db: %w", err)
 			}
-			errbuf := bytes.Buffer{}
-			cmd.Stderr = &errbuf
-			zlog.Debug(ctx).Strs("cmd", cmd.Args).Msg("rpm invocation")
-			if err := cmd.Start(); err != nil {
-				r.Close()
-				return nil, err
+			defer func() {
+				if err := r.Close(); err != nil {
+					zlog.Warn(ctx).Err(err).Msg("unable to close tarfs sqlite db")
+				}
+			}()
+			f, err := os.CreateTemp(os.TempDir(), `rpmdb.sqlite.*`)
+			if err != nil {
+				return nil, fmt.Errorf("rpm: error reading sqlite db: %w", err)
 			}
-			// Use a closure to defer the Close call.
-			if err := func() error {
-				defer r.Close()
-				srcs := make(map[string]*claircore.Package)
-				s := bufio.NewScanner(r)
-				s.Split(querySplit)
-
-				for s.Scan() {
-					p, err := parsePackage(ctx, srcs, bytes.NewBuffer(s.Bytes()))
-					if err != nil {
-						return err
-					}
-					p.PackageDB = db.String()
-					pkgs = append(pkgs, p)
+			defer func() {
+				if err := os.Remove(f.Name()); err != nil {
+					zlog.Error(ctx).Err(err).Msg("unable to unlink sqlite db")
 				}
-
-				return s.Err()
-			}(); err != nil {
-				if errbuf.Len() != 0 {
-					zlog.Warn(ctx).
-						Strs("cmd", cmd.Args).
-						Str("err", errbuf.String()).
-						Msg("error output")
+				if err := f.Close(); err != nil {
+					zlog.Warn(ctx).Err(err).Msg("unable to close sqlite db")
 				}
-				return nil, fmt.Errorf("rpm: error reading rpm output: %w", err)
+			}()
+			zlog.Debug(ctx).Str("file", f.Name()).Msg("copying sqlite db out of tar")
+			if _, err := io.Copy(f, r); err != nil {
+				return nil, fmt.Errorf("rpm: error reading sqlite db: %w", err)
+			}
+			if err := f.Sync(); err != nil {
+				return nil, fmt.Errorf("rpm: error reading sqlite db: %w", err)
 			}
-			if err := cmd.Wait(); err != nil {
-				return nil, err
+			sdb, err := sqlite.Open(f.Name())
+			if err != nil {
+				return nil, fmt.Errorf("rpm: error reading sqlite db: %w", err)
 			}
-		case kindSQLite, kindNDB:
-			var nat nativeDB // see native_db.go:/nativeDB
-			switch db.Kind {
-			case kindSQLite:
-				r, err := sys.Open(path.Join(db.Path, `rpmdb.sqlite`))
+			defer sdb.Close()
+			nat = sdb
+		case kindBDB:
```

Comment:
Seems like this could be compressed further given the procedure is almost identical for `bdb` and `ndb` but could affect readability